### ISSUE TITLE
MAUI iOS error on version 1.5.0 - System.DllNotFoundException

### ISSUE THREAD

**Issue by kaniosm** [sandrohanea/whisper.net#178]
I'm getting the below error with version 1.5.0 on MAUI running on iOS

System.DllNotFoundException: __Internal
   at Whisper.net.Logger.LogProvider.InitializeLogging()
   at Whisper.net.WhisperFactory.<>c.<.cctor>b__11_0()
   at System.Lazy`1[[Whisper.net.LibraryLoader.LoadResult, Whisper.net, Version=1.5.0.0,

**Comment by sandrohanea** [comment]
Hello @kaniosm ,
Did you managed to fix the issue?

Can you, please, try again with version 1.7.0 and if the issue still occur to let us know what is your target framework? It should be supported for Maui as long as [Whisper.net.Runtime](https://www.nuget.org/packages/Whisper.net.Runtime) is installed and minimum target framework version is net8.0-ios

**Comment by kaniosm** [comment]
Hi @sandrohanea ,

I still get the same error (see below).
The project is .net 8.0 (MAUI 8.0.91) and I'm deploy to my tablet (ipad 8th gen) running tabOS 18.

{System.DllNotFoundException: __Internal
   at Whisper.net.Logger.LogProvider.InitializeLogging()
   at Whisper.net.WhisperFactory.<>c.<.cctor>b__9_0()
   at System.Lazy`1[[Whisper.net.LibraryLoader.LoadResult, Whisper.net, Version=1.7.0.0, Culture=neutral, PublicKeyToken=null]].ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1[[Whisper.net.LibraryLoader.LoadResult, Whisper.net, Version=1.7.0.0, Culture=neutral, PublicKeyToken=null]].ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1[[Whisper.net.LibraryLoader.LoadResult, Whisper.net, Version=1.7.0.0, Culture=neutral, PublicKeyToken=null]].CreateValue()
   at System.Lazy`1[[Whisper.net.LibraryLoader.LoadResult, Whisper.net, Version=1.7.0.0, Culture=neutral, PublicKeyToken=null]].get_Value()
   at Whisper.net.WhisperFactory..ctor(IWhisperProcessorModelLoader loader, Boolean delayInit)
   at Whisper.net.WhisperFactory.FromPath(String path, Boolean delayInitialization)

**Comment by sandrohanea** [comment]
Unfortunately, I don't know why this can be and I don't have a Mac anymore to test it. 
Here are the runtime artefacts for ios: [runtime.zip](https://github.com/user-attachments/files/17328452/runtime.zip)

Can you, please, try to include the runtime dylibs manually in your project (root of the project) with `CopyIfNewer`:
![image](https://github.com/user-attachments/assets/c12664bf-7d64-40dd-99b4-b38613f19607)
![image](https://github.com/user-attachments/assets/d198fd11-6d21-4ad3-9abc-7d24a4e1f1c9)

You can also remove `Whisper.net.Runtime` dependency.
This way we can confirm if the issue is with the Whisper.net.Runtime or with the dylib itself.

**Comment by sandrohanea** [comment]
You can also watch: https://github.com/sandrohanea/whisper.net/issues/212 where it seems someone else had a similar issue and managed to fix it by adding the files manually.

**Comment by kaniosm** [comment]
Thanks for the update @sandrohanea .

I'll give it a try some time when I get the chance. At the moment this is not critical as we were experimenting with Whisper to understand if we can have on-device transcriptions....
At the moment we do this over a background service.
I had a similar issue with wihsper.net running on linux, where I had to copy the .so files from the runtime folder to the application folder.

**Comment by AncientLust** [comment]
Hi @sandrohanea, thank you to your team for this project.

Having the same issue with iOS in MAUI project.

1. Could you please clarify why it was decided to use `.dylib` libraries for iOS? As far as I know dynamic libraries outside of a framework bundle, are not supported on iOS [source](https://stackoverflow.com/a/75101594/22742076).

2. Sorry for probably strange question, but was these `.dylib` libraries ever tested on an iOS device inside a .NET app?

**Comment by AncientLust** [comment]
Hi @sandrohanea, thank you and your team for this work.

Having the same issue with iOS in MAUI project.

1. Could you please clarify why it was decided to use `.dylib` libraries for iOS? As far as I know dynamic libraries outside of a framework bundle are not supported on iOS [source 1](https://stackoverflow.com/a/75101594/22742076) and [source 2](https://forums.developer.apple.com/forums/thread/709854?utm_source=chatgpt.com) . I believe static libraries `.a` should be used instead. For example Vosk provides `.a` libraries (for enterprise only) which work seamlesly via PINVOKE in MAUI iOS.

2. Sorry for probably strange question, but was these `.dylib` libraries ever tested on an iOS device inside a .NET app?

> It should be supported for Maui as long as [Whisper.net.Runtime](https://www.nuget.org/packages/Whisper.net.Runtime) is installed and minimum target framework version is net8.0-ios

**Comment by sandrohanea** [comment]
Hello @AncientLust ,

For question 1:
Initial support for iOS apps was added in https://github.com/sandrohanea/whisper.net/commit/10cd30c8f7d0843a80782ae606ef86df4d15904e by @drasticactions using dylibs and `_Internal` marked DllImports (and it was working at the time).

Also, whisper.cpp is building dynamic libraries by default but it needs to be investigated if we can provide static libraries instead and if that will fix the issues.

For question 2:
This library is indeed not tested for a couple of versions on ios-simulator / ios-devices as I don't own a mac anymore (this is why this issue is still opened together with the rest of the issues that requires a mac to test (e.g. https://github.com/sandrohanea/whisper.net/issues/212))

Contributions are welcomed, but for now, I don't have the possibility to fix and test the issues for macos / ios / tvos.

**Comment by AncientLust** [comment]
Thanks for a quick reply @sandrohanea.

> Also, whisper.cpp is building dynamic libraries by default but it needs to be investigated if we can provide static libraries instead and if that will fix the issues.

1. Are you sure? As far as I understand whisper.cpp provides cpp sources. While this repo builds these sources into platform-specific libraries via `MakeFile`. Adjusting cmake scripts I was able to build `.a` libraries, but still didn't make it working. I'll get back if I have success with it.

> This library is indeed not tested for a couple of versions on ios-simulator / ios-devices as I don't own a mac anymore (this is why this issue is still opened together with the rest of the issues that requires a mac to test (e.g. https://github.com/sandrohanea/whisper.net/issues/212))

2. Could you please share version of Whisper.net which was smoke-tested on iOS? So far I also tested the old one 1.4 which shows the same issue. It might help to investigate what's wrong. I believe `.dylib` also should work (not sure), but high chances it will bring issues when App Store publication comes to place.

3. Do I have a right understanding that both `Whisper.net.Runtime` and `Whisper.net.Runtime.CoreML` supposed to work in MAUI iOS? Looking into NuGet description [here](https://www.nuget.org/packages/Whisper.net.Runtime/) it should be true.

4. Could it be an issue that I'm using Mac for `Whisper.net.Runtime` and it's marked as `macOS: TBD`? Could you please elaborate what state means `TBD`? Does `Whisper.net.Runtime.CoreML` has the same 'state'? Because it doesn't have section `Pre-requisites` which could answer this question, like other runtimes.

**Comment by AncientLust** [comment]
Thanks for a quick reply @sandrohanea.

> Also, whisper.cpp is building dynamic libraries by default but it needs to be investigated if we can provide static libraries instead and if that will fix the issues.

1. Are you sure? As far as I understand whisper.cpp provides cpp sources. While this repo builds these sources into platform-specific libraries via `MakeFile`. Adjusting cmake scripts I was able to build `.a` libraries, but still didn't make it work. I'll get back if I have success with it.

> This library is indeed not tested for a couple of versions on ios-simulator / ios-devices as I don't own a mac anymore (this is why this issue is still opened together with the rest of the issues that requires a mac to test (e.g. https://github.com/sandrohanea/whisper.net/issues/212))

2. Could you please share version of Whisper.net which was smoke-tested on iOS? So far I also tested the old one 1.4 which shows the same issue. It might help to investigate what's wrong. I believe `.dylib` also should work (not sure), but high chances it will bring issues when App Store publication comes to place.

3. Do I have a right understanding that both `Whisper.net.Runtime` and `Whisper.net.Runtime.CoreML` supposed to work in MAUI iOS? Looking into NuGet description [here](https://www.nuget.org/packages/Whisper.net.Runtime/) it should be true.

4. Could it be an issue that I'm using Mac for `Whisper.net.Runtime` and it's marked as `macOS: TBD`? Could you please elaborate what state means `TBD`? Does `Whisper.net.Runtime.CoreML` has the same 'state'? Because it doesn't have section `Pre-requisites` which could answer this question, like other runtimes.

**Comment by sandrohanea** [comment]
1. Yes, we're building the libraries using the Makefile and github actions. I just mentined that default configuration will result in dylibs (but ofc we can change that if required)

2. 1.4.0 was the first version after the support for ios was added. Since then, there was small or non-existing testing for ios.

3. Yes, they should work, but as mentioned, this is not tested and it is highly likely to contain bugs). They are bundling the native libraries and using dll import with `_Internal` dll name to use the native dlls.

4. When I updated the readme to include the dependency, I didn't know exactly what are the minimum requrments for macos (as I knew for Windows and Linux) so this is why I left it "to be defined". I know that library works on macos as I have github actions and tests to check it, but cannot test too many setups / options. There are also tests for Android, Linux and Windows but there are not tests for Ios / TVOS / macCatalyst currently.

**Comment by AncientLust** [comment]
> Yes, they should work, but as mentioned, this is not tested and it is highly likely to contain bugs). They are bundling the native libraries and using dll import with _Internal dll name to use the native dlls.

You've mentioned `_Internal` with one underscore. I'm looking into PR description from [here](https://github.com/sandrohanea/whisper.net/pull/30) and see that @drasticactions mentioned `__Internal` with two underscores. I think that’s the thing. 

In my understanding `__Internal` is reserved for static linking, while `_Internal` is probably not valid because we don't have libraries with name `_Internal`. And looking into `Whisper.net/Internals/Native/Implementations` I don't see any mention of `__Internal`.

@drasticactions could you please drop a comment if that's correct?

**Comment by AncientLust** [comment]
Yep, I can confirm that changing `_Internal` to `__Internal` in `LibraryImportInternalWhisper` solves the issue with `System.DllNotFoundException: _Internal`. Just recognized `wav` file on iOS.

@sandrohanea, I can create PR, but I have one question. Aside the change with `__Internal` I also had to uncomment `USE_WHISPER_MAUI` in `Whisper.net.csproj` in order to make it work with my MAUI test proj:

```csharp
// Uncommented this section
<PropertyGroup>
  <USE_WHISPER_MAUI>TRUE</USE_WHISPER_MAUI>
</PropertyGroup>

// Dropped android, tv-ios and maccatalyst since my test project targeted only net8.0-ios
<PropertyGroup Condition="$(USE_WHISPER_MAUI) != ''">
  <TargetFrameworks>
    net8.0;net6.0;netstandard2.0;net8.0-ios
  </TargetFrameworks>
</PropertyGroup>
```

Without this change `NativeLibraryLoader` didn't fall into `IOS || MACCATALYST || TVOS` directive and tried to load library dynamically via `LoadLibraryComponent()`.

Could you please give a comment why it was commented out? I have assumption that when project packed into a NuGet and loaded into MAUI this directive is not necessary and kept for testing only, like in my case.

**Comment by sandrohanea** [comment]
Hey @AncientLust ,

That's great news! Indeed, it seems that I introduced the issue in one of my refactorings and forgot to add an extra underscore. 

That line in the Whisper.csproj is just to set that variable, but indeed, the variable is set as an environment variable in the github workflow that is producing the nuget package artefacts.

The reason why I added that config with the variable is to allow people without maui workloads installed to experiment with the app and improve build times for the rest of the workflows (e.g unit tests for linux).

Thanks for taking the time to investigate and for the fix!

Feel free to create a PR if you want and will release it with the next version (preparing some version that upgrades whisper and adds support for net9 with maui).
